transport/awssqs/consumer.go

```diff
@@ -0,0 +1,296 @@
+// visibilityTimeout field.
+func ConsumerVisibilityTimeout(visibilityTimeout int64) ConsumerOption {
+	return func(c *Consumer) { c.visibilityTimeout = visibilityTimeout }
+}
+
+// ConsumerWantReplyFunc overrides the default value for the consumer's
+// wantRep field.
+func ConsumerWantReplyFunc(replyFunc WantReplyFunc) ConsumerOption {
+	return func(c *Consumer) { c.wantRep = replyFunc }
+}
+
+// ConsumerErrorHandler is used to handle non-terminal errors. By default, non-terminal errors
+// are ignored. This is intended as a diagnostic measure. Finer-grained control
+// of error handling, including logging in more detail, should be performed in a
+// custom ConsumerErrorEncoder which has access to the context.
+func ConsumerErrorHandler(errorHandler transport.ErrorHandler) ConsumerOption {
+	return func(c *Consumer) { c.errorHandler = errorHandler }
+}
+
+// ConsumerFinalizer is executed once all the received SQS messages are done being processed.
+// By default, no finalizer is registered.
+func ConsumerFinalizer(f ...ConsumerFinalizerFunc) ConsumerOption {
+	return func(c *Consumer) { c.finalizer = f }
+}
+
+// ConsumerDeleteMessage overrides the default value for the consumer's
+// deleteMessage field to indicate when the consumed messages should be deleted.
+func ConsumerDeleteMessage(delete Delete) ConsumerOption {
+	return func(c *Consumer) { c.deleteMessage = delete }
+}
+
+// Consume calls ReceiveMessageWithContext and handles messages having an
+// sqs.ReceiveMessageInput as parameter allows each user to have his own receive configuration.
+// That said, this method overrides the queueURL for the provided ReceiveMessageInput to ensure
+// the messages are retrieved from the consumer's configured queue.
+func (c Consumer) Consume(ctx context.Context, receiveMsgInput *sqs.ReceiveMessageInput) error {
+	receiveMsgInput.QueueUrl = &c.queueURL
+	out, err := c.sqsClient.ReceiveMessageWithContext(ctx, receiveMsgInput)
+	if err != nil {
+		return err
+	}
+	return c.handleMessages(ctx, out.Messages)
+}
+
+// handleMessages handles the consumed messages.
+func (c Consumer) handleMessages(ctx context.Context, msgs []*sqs.Message) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Copy received messages slice in leftMsgs slice
+	// leftMsgs will be used by the consumer's visibilityTimeoutFunc to extend the
+	// visibility timeout for the messages that have not been processed yet.
+	leftMsgs := []*sqs.Message{}
+	leftMsgs = append(leftMsgs, msgs...)
+
+	visibilityTimeoutCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go c.visibilityTimeoutFunc(visibilityTimeoutCtx, c.sqsClient, c.queueURL, c.visibilityTimeout, &leftMsgs, c.leftMsgsMux)
+
+	if len(c.finalizer) > 0 {
+		defer func() {
+			for _, f := range c.finalizer {
+				f(ctx, &msgs)
+			}
+		}()
+	}
+
+	for _, f := range c.before {
+		ctx = f(ctx, &msgs)
+	}
+
+	for _, msg := range msgs {
+		if c.deleteMessage == BeforeHandle {
```

Comment:
this process delete or after handle message can be done inside before or after function, right?